src/serac/physics/utilities/variational_form/weak_form.hpp

```diff
@@ -0,0 +1,500 @@
+  const bool                 is_segment = fe->GetGeomType() == mfem::Geometry::SEGMENT;
+  const bool                 is_hcurl   = fe->GetMapType() == mfem::FiniteElement::MapType::H_CURL;
+  if (is_segment && is_hcurl) {
+    return mfem::ElementDofOrdering::NATIVE;
+  } else {
+    return mfem::ElementDofOrdering::LEXICOGRAPHIC;
+  }
+}
+#endif
+
+/// @cond
+template <typename T>
+class WeakForm;
+/// @endcond
+
+/**
+ * @brief Intended to be like @p std::function for finite element kernels
+ *
+ * That is: you tell it the inputs (trial spaces) for a kernel, and the outputs (test space) like @p std::function
+ * For example, this code represents a function that takes an integer argument and returns a double:
+ * @code{.cpp}
+ * std::function< double(double, int) > my_func;
+ * @endcode
+ * And this represents a function that takes values from an Hcurl field and returns a
+ * residual vector associated with an H1 field:
+ * @code{.cpp}
+ * WeakForm< H1(Hcurl) > my_residual;
+ * @endcode
+ *
+ * @tparam test The space of test functions to use
+ * @tparam trial The space of trial functions to use
+ *
+ * To use this class, you use the methods @p WeakForm::Add****Integral(integrand,domain_of_integration)
+ * where @p integrand is a q-function lambda or functor and @p domain_of_integration is an @p mfem::mesh
+ *
+ * @see https://libceed.readthedocs.io/en/latest/libCEEDapi/#theoretical-framework for additional
+ * information on the idea behind a quadrature function and its inputs/outputs
+ *
+ * @code{.cpp}
+ * // for domains made up of quadrilaterals embedded in R^2
+ * my_residual.AddAreaIntegral(integrand, domain_of_integration);
+ * // alternatively...
+ * my_residual.AddDomainIntegral(Dimension<2>{}, integrand, domain_of_integration);
+ *
+ * // for domains made up of quadrilaterals embedded in R^3
+ * my_residual.AddSurfaceIntegral(integrand, domain_of_integration);
+ *
+ * // for domains made up of hexahedra embedded in R^3
+ * my_residual.AddVolumeIntegral(integrand, domain_of_integration);
+ * // alternatively...
+ * my_residual.AddDomainIntegral(Dimension<3>{}, integrand, domain_of_integration);
+ * @endcode
+ */
+template <typename test, typename trial>
+class WeakForm<test(trial)> : public mfem::Operator {
+public:
+  /**
+   * @brief Constructs using @p mfem::ParFiniteElementSpace objects corresponding to the test/trial spaces
+   * @param[in] test_fes The test space
+   * @param[in] trial_fes The trial space
+   */
+  WeakForm(mfem::ParFiniteElementSpace* test_fes, mfem::ParFiniteElementSpace* trial_fes)
+      : Operator(test_fes->GetTrueVSize(), trial_fes->GetTrueVSize()),
+        test_space_(test_fes),
+        trial_space_(trial_fes),
+        P_test_(test_space_->GetProlongationMatrix()),
+        G_test_(test_space_->GetElementRestriction(mfem::ElementDofOrdering::LEXICOGRAPHIC)),
+        // G_test_boundary_(test_space_->GetFaceRestriction(get_boundary_dof_ordering(test_fes),
```

Comment:
I think we should punt on the boundary integrator stuff for now until the needed upstream MFEM fix happens. A comment reflecting this is a good idea.